File: feign-core/src/main/java/name/abhijitsarkar/feign/core/web/RequestHandlerMethodArgumentResolver.java
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * License for more details.
 */

package name.abhijitsarkar.feign.core.web;

import name.abhijitsarkar.feign.Request;
import org.springframework.core.MethodParameter;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

import javax.servlet.http.HttpServletRequest;
import java.util.Map;

import static java.lang.System.lineSeparator;
import static java.util.Collections.list;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toMap;

/**
 * @author Abhijit Sarkar
 */
public class RequestHandlerMethodArgumentResolver implements HandlerMethodArgumentResolver {
    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return Request.class.isAssignableFrom(parameter.getParameterType());
    }

    @Override
    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer,
                                  NativeWebRequest webRequest, WebDataBinderFactory binderFactory) throws Exception {
        HttpServletRequest request = (HttpServletRequest) webRequest.getNativeRequest();

        Map<String, String> headers = list(request.getHeaderNames()).stream()
                .collect(toMap(name -> name, request::getHeader));

        String body = request.getReader().lines().collect(joining(lineSeparator()));

        return Request.builder()
                .path(request.getServletPath())
                .method(request.getMethod())
                .queryParams(request.getParameterMap())
                .headers(headers)
                .body(body)
                .build();
    }
}

File: feign-core/src/main/scala/name/abhijitsarkar/feign/core/web/RequestHandlerMethodArgumentResolver.scala
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of the GNU General Public License accompanies this software,
 * and is also available at http://www.gnu.org/licenses.
 */

package name.abhijitsarkar.feign.core.web

import javax.servlet.http.HttpServletRequest

import name.abhijitsarkar.feign.Request
import org.springframework.core.MethodParameter
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.{HandlerMethodArgumentResolver, ModelAndViewContainer}

import scala.collection.JavaConverters._
import scala.collection.immutable.{Map => ImmutableMap}

/**
  * @author Abhijit Sarkar
  */
class RequestHandlerMethodArgumentResolver extends HandlerMethodArgumentResolver {
  override def resolveArgument(parameter: MethodParameter, mavContainer: ModelAndViewContainer,
                               webRequest: NativeWebRequest, binderFactory: WebDataBinderFactory): Request = {
    val request = webRequest.getNativeRequest.asInstanceOf[HttpServletRequest]

    val headers = request.getHeaderNames.asScala.foldLeft(ImmutableMap[String, String]()) { (map, h) =>
      map + (h -> request.getHeader(h))
    }

    val br = request.getReader
    val body = Stream.continually(br.readLine()).takeWhile(_ != null).mkString("\n")

    val req = new Request()
    req.path = request.getServletPath
    req.method = request.getMethod
    req.queryParams = request.getParameterMap.asScala.mapValues(_.toList).toMap
    req.headers = headers
    req.body = Option(body)

    req
  }

  override def supportsParameter(parameter: MethodParameter): Boolean =
    classOf[Request].isAssignableFrom(parameter.getParameterType)
}
